internal/merkle/inmemory/reference_test.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2022 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inmemory
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/bits"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/transparency-dev/merkle"
+	"github.com/transparency-dev/merkle/rfc6962"
+	to "github.com/transparency-dev/merkle/testonly"
+)
+
+// The reference Merkle tree hashing and proof algorithms in this file directly
+// implement the definitions from RFC 6962 [1]. We use this implementation only
+// for testing correctness of other more flexible and performant algorithms,
+// such as the in-memory Tree type and compact ranges.
+//
+// [1] https://datatracker.ietf.org/doc/html/rfc6962#section-2
+
+// refRootHash returns the root hash of a Merkle tree with the given entries.
+// This is a reference implementation for cross-checking.
+func refRootHash(entries [][]byte, hasher merkle.LogHasher) []byte {
```

Comment:
Do the functions here need to be prepended with "ref"? IIUC, "reference" here refers to the fact that this is returns an example/specific case for a root hash. However, there is nothing specific about this code, since it takes `entries` as a param. Thus, is it inaccurate to just call it `rootHash`?

Reply:
"reference" here mean that the implementation is reference (not specific values/hashes). See the comment in lines 29-34.

Reply:
To test various Merkle tree implementation, there are a couple of techniques:
- "golden" tests: use specific inputs and outputs to check that it matches exactly. Pros: no way to corrupt the "expected" result, cons: need to consider lots of corner cases manually.
- "reference" implementation: compare our implementation to ones that are most simple/intuitive (but maybe less performanct) and follow definitions closely; we know that these implementations probably work, so we can run an exhaustive test that verifies new algorithm on all possible inputs.

Reply:
So, the methodology is:
- Make reference implementations (this file).
- Test them against golden tests (this file).
- Test any other implementation exhaustively against the reference implementation (other files).
  - Note: Other tests will indirectly "inherit" all the golden tests, so there is no need to repeat the golden testing there.

Reply:
This way, we get Pros of both approaches, and minimise Cons.

Reply:
This also partly addresses you suggestion below of exposing the "known good" values. It can be done, but is not strictly necessary in the described methodology.

Reply:
Gotcha, thanks. Re-reading your comments in the code now makes a lot more sense to me.